sdk/core/System.ClientModel/src/Internal/Logging/PipelineMessageLogger.cs

```diff
@@ -0,0 +1,258 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace System.ClientModel.Internal;
+
+internal partial class PipelineMessageLogger
+{
+    private readonly ILogger<MessageLoggingPolicy>? _logger;
+    private readonly PipelineMessageSanitizer _sanitizer;
+
+    public PipelineMessageLogger(PipelineMessageSanitizer sanitizer, ILoggerFactory? loggerFactory)
+    {
+        _sanitizer = sanitizer;
+        _logger = loggerFactory?.CreateLogger<MessageLoggingPolicy>() ?? null;
+    }
+
+    /// <summary>
+    ///  Whether the given log level or event level is enabled, depending
+    ///  on whether this handler logs to ILogger or Event Source. Can be
+    ///  used to guard expensive operations.
+    /// </summary>
+    /// <param name="logLevel"></param>
+    /// <param name="eventLevel"></param>
+    public bool IsEnabled(LogLevel logLevel, EventLevel eventLevel)
+    {
+        return _logger is not null ? _logger.IsEnabled(logLevel) : ClientModelEventSource.Log.IsEnabled(eventLevel, EventKeywords.None);
+    }
+
+    #region Request
+
+    public void LogRequest(string requestId, PipelineRequest request, string? clientAssembly)
+    {
+        if (_logger is not null)
+        {
+            if (_logger.IsEnabled(LogLevel.Information))
+            {
+                Request(_logger, requestId, request.Method, _sanitizer.SanitizeUrl(request.Uri!.AbsoluteUri), FormatHeaders(request.Headers), clientAssembly);
+            }
+        }
+        else
+        {
+            ClientModelEventSource.Log.Request(requestId, request, clientAssembly, _sanitizer);
+        }
+    }
+
+    [LoggerMessage(LoggingEventIds.RequestEvent, LogLevel.Information, "Request [{requestId}] {method} {uri}\r\n{headers}client assembly: {clientAssembly}", SkipEnabledCheck = true, EventName = "Request")]
+    private static partial void Request(ILogger logger, string requestId, string method, string uri, string headers, string? clientAssembly);
```

Comment:
The source generated partial method gets confused and produces compile warnings about potential null dereferencing if the ILogger property is nullable, so changed these to be static and then passed in the ILogger directly after checking it's not null

Reply:
Sounds like a good use of nullable annotations to prevent a `NullReferenceException`!